rectools/metrics/auc.py

```diff
@@ -0,0 +1,512 @@
+        # We calculate and keep number of false positives for all other users
+        n_fp_insufficient = users_n_fp[users_n_fp < k_max]
+        users_with_fn = outer_merged.loc[outer_merged[Columns.Rank].isna(), Columns.User].unique()
+        n_fp_insufficient = n_fp_insufficient[n_fp_insufficient.index.isin(users_with_fn)]
+
+        return AUCFitted(outer_merged, n_pos, n_fp_insufficient)
+
+    def _get_sufficient_reco_explanation(self) -> str:
+        raise NotImplementedError()
+
+    def _handle_insufficient_cases(
+        self, outer_merged: pd.DataFrame, n_pos: pd.Series, n_fp_insufficient: pd.Series, metric_name: str
+    ) -> pd.Series:
+        if self.insufficient_handling == InsufficientHandling.IGNORE:
+            return outer_merged, n_pos
+
+        insufficient_users = n_fp_insufficient[n_fp_insufficient < self.k].index.values
+        if not insufficient_users.any():
+            return outer_merged, n_pos
+
+        if self.insufficient_handling == InsufficientHandling.EXCLUDE:
+            outer_merged_suf = outer_merged[~outer_merged[Columns.User].isin(insufficient_users)]
+            n_pos_suf = n_pos[~n_pos.index.isin(insufficient_users)]
+            return outer_merged_suf, n_pos_suf
+
+        raise ValueError(
+            f"""
+            {metric_name}@{self.k} metric requires at least {self.k} negatives in
+            recommendations for each user. Or all items from user test interactions ranked in
+            recommendations - meaning that all other recommended items will be negatives.
+            There are {len(insufficient_users)} users with less then required negatives.
+            For correct {metric_name} computation please provide each user with sufficient number
+            of recommended items. {self._get_sufficient_reco_explanation()}
+            You can disable this error by specifying `insufficient_handling`="{InsufficientHandling.IGNORE}" or
+            by excluding all users with insuffissient recommendations from metric computation
+            with specifying `insufficient_handling` = "{InsufficientHandling.EXCLUDE}".
+            """
+        )
+
+    def _calc_roc_auc(self, cropped_outer_merged: pd.DataFrame, n_pos: pd.Series) -> pd.Series:
+        """
+        Calculate ROC AUC given that all data has already been prepared, merged, enriched and cropped following
+        metric specific logic.
+        """
+        auc_numenator_gain = (self.k - cropped_outer_merged["__fp_cumsum"]) * cropped_outer_merged["__tp"]
+        auc_numenator_gain.name = "__auc_numenator_gain"
+        user_auc_numenator = pd.concat([cropped_outer_merged[Columns.User], auc_numenator_gain], axis=1)
+        auc_numenator = user_auc_numenator.groupby(Columns.User)["__auc_numenator_gain"].sum()
+        auc_denominator = n_pos * self.k
+        auc = (auc_numenator / auc_denominator).fillna(0)  # TODO: what if order was changed?
```

Comment:
Good question. It's better to merge them IMO

Reply:
fixed